CounterApp/App/AppCore.swift

```diff
@@ -0,0 +1,55 @@
+//
+//  AppCore.swift
+//  CounterApp
+//
+//  Created by Wenjuan Li on 2022/7/12.
+//
+
+import ComposableArchitecture
+import Foundation
+
+struct AppState: Equatable {
+    var counter: CounterState
+    var lock: LockState
+}
+
+enum AppAction: Equatable {
+    case counter(CounterAction)
+    case lock(LockAction)
+}
+
+struct AppEnviroment {}
+
+let appReducer = Reducer<AppState, AppAction, AppEnviroment>.combine(
+    counterReducer.pullback(
+        state: \AppState.counter,
+        action: /AppAction.counter,
+        environment: { _ in CounterEnviroment.defaultEnv() }
```

Comment:
Generally, local domain environments are also scope down from global domain ones. `AppEnvironment` should hold all the dependencies needed to be able to build the `CounterEnvironment`. For sake of readability, is also good to define an extension over the global domain environment, this case it would be `AppEnvironment`, that exposes a property that returns the local domain environment, in this case `CounterEnvironment`, so you could satisfy the (GlobalEnvironment) -> LocalEnvironment closure from the pullback method by using its key path. Then the pullback would look more like this:
```
counterReducer.pullback(
     state: \.counter,
     action: /AppAction.counter,
     environment: \.counter
)
```

Reply:
@nanocba Fixed